### PR TITLE
[aptos-rest-client / rosetta / cli] Rest client gas estimation

### PR DESCRIPTION
**Pull request by gregnazario**
### Description
Gas estimation is now common between rosetta, and the CLI.  Handles minimum and maximum gas properly.  Will remove defaults once this is properly confirmed.

### Test Plan
E2E tests

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/4273)

### CODE REVIEW

File: api/src/transactions.rs [hunk 1]
```diff
@@ -360,7 +368,74 @@ impl TransactionsApi {
...
+                .ok_or_else(|| {
+                    SubmitTransactionError::bad_request_with_code(
+                        "Account not found",
+                        AptosErrorCode::InvalidInput,
+                        &ledger_info,
+                    )
+                })?;
+            let coin_store: CoinStoreResource = account_state
+                .get_coin_store_resource()
+                .and_then(|inner| inner.ok_or_else(|| anyhow!("No coin store found!")))
```
Comment (banool) [inline api/src/transactions.rs:402]: Maybe `"no count store found on account {}", address` just to be extra clear.

File: api/src/transactions.rs [hunk 2]
```diff
@@ -360,7 +368,74 @@ impl TransactionsApi {
...
+                        format!("Failed to get coin store resource {}", err),
+                        AptosErrorCode::InternalError,
+                        &ledger_info,
+                    )
+                })?;
+
+            let gas_unit_price =
+                estimated_gas_unit_price.unwrap_or_else(|| signed_transaction.gas_unit_price());
+
+            // With 0 gas price, we set it to max gas units, since we can't divide by 0
```
Comment (banool) [inline api/src/transactions.rs:414]: lol nice

File: crates/aptos/src/common/types.rs [hunk 3]
```diff
@@ -1239,23 +1238,37 @@ impl TransactionOptions {
...
-                    "Simulated transaction failed with status {}",
-                    simulated_txn.info.vm_status
-                )));
-            }
-            simulated_txn.info.gas_used.0
+            simulated_txn
+                .into_inner()
+                .transaction
+                .as_signed_user_txn()
+                .expect("Should be a user transaction from simulation!")
```
Comment (banool) [inline crates/aptos/src/common/types.rs:1270]: This func returns a Result, use `.context` instead? nbd since this should never happen though.

**Review by banool — APPROVED** [review]
Looks solid. All the stuff seems to be regenerated correctly too.

Since you've bumped the API version, could you add a changelog entry. If you want to let SDK users know about this, add an entry in the "unreleased" section in the SDK changelog too.

File: api/src/transactions.rs [hunk 4]
```diff
@@ -360,7 +368,74 @@ impl TransactionsApi {
...
-        let signed_transaction = self.get_signed_transaction(&ledger_info, data)?;
+        let mut signed_transaction = self.get_signed_transaction(&ledger_info, data)?;
+
+        let estimated_gas_unit_price = if estimate_gas_unit_price.0.unwrap_or_default() {
+            Some(self.context.estimate_gas_price(&ledger_info)?)
+        } else {
+            None
+        };
+
+        // If estimate max gas amount is provided, we will just make it the maximum value
```
Comment (movekevin) [inline api/src/transactions.rs:379]: You mean "If estimate max gas amount is NOT provided"?

File: api/src/transactions.rs [hunk 5]
```diff
@@ -360,7 +368,74 @@ impl TransactionsApi {
...
+                })?;
+
+            let gas_unit_price =
+                estimated_gas_unit_price.unwrap_or_else(|| signed_transaction.gas_unit_price());
+
+            // With 0 gas price, we set it to max gas units, since we can't divide by 0
+            let max_account_gas_units = if gas_unit_price == 0 {
+                coin_store.coin()
+            } else {
+                coin_store.coin() / gas_unit_price
```
Comment (movekevin) [inline api/src/transactions.rs:418]: Can we not do division and instead compare using multiplication (+ even casting to u128 if needed)? Multiplication can lead to off by 1 errors due to rounding

File: api/src/transactions.rs (hunk 4 above)
Comment (movekevin) [reply api/src/transactions.rs:379]: Oh it's bool. Nvm.

**Review by movekevin — APPROVED** [review]

File: api/src/transactions.rs (hunk 5 above)
Comment (gregnazario) [reply api/src/transactions.rs:418]: This handles the following case:

I have a global max amount of gas units of 3.

The gas price is 10.

If I have 20-29 coins, I can only pay for gas 2 gas units (e.g. 29 / 10 rounded down).

I compare 2 to 3, and I have the minimum that is supposed to be 2 and not 3.

If I have 39 coins, I can only pay for 3 gas units (e.g. 39 / 10 rounded down).

I compare 3 to 3 and I have the minimum of that which is 3.

If I have 40 coins, I can only pay for 4 gas units, but the max number I can have is 3.

Min(4, 3) = 3

**Pull request merged**